draft-normington-jsonpath-latest.xml

```diff
@@ -218,6 +225,94 @@ dot-child-name = TBD
+                     </t>
+                     <sourcecode type="abnf">
+union-element = union-child ; see below for more alternatives
+union-child = %x27 *single-quoted %x27 / ; 'string'
+              %x22 *double-quoted %x22   ; "string"
+
+single-quoted = %x5C %x27 / ; \' (escaped single quote)
+                %x5C %x5C / ; \\ (escaped reverse solidus)
+                %x00-26 / %x28-5B %x5D-D7FF / %xE000-10FFFF
+                            ; any Unicode code point except ' or \
+
+double-quoted = %x5C 5x22 / ; \" (escaped double quote)
+                %x5C 5x5C / ; \\ (escaped reverse solidus)
+                %x00-21 / %x23-5B / %x5D-D7FF / %xE000-10FFFF
+                            ; any Unicode code point except " or \
+                     </sourcecode>
+                     <t>Note: the syntax does not allow any character other than a quote or reverse solidus to be escaped.
+                     </t>
+                  </section>       
+                  <section numbered="false" toc="exclude">
+                     <name>Semantics</name>
+                     <t>If the union child is a quoted string, the string must first have the surrounding quotes removed and then be unescaped
+                        as describe below.
+                        The union child matches any node which is an object having a key equal to the unescaped value of the union child string
+                        and selects the value corresponding to that key.
+                        It does not match a node which is not an object.
+                     </t>
+                     <t>After the surrounding quotes have been removed, a string is unescaped by replacing each escaped quote by the corresponding
+                        quote and replacing each escaped reverse solidus by a reverse solidus.
+                     </t>
+                  </section>       
+               </section>             
+               <section numbered="true" toc="default">
+                  <name>Array Index</name>
+                  <section numbered="false" toc="exclude">
+                     <name>Syntax</name>
+                     <t>An array index is a union element consisting of an integer (in base 10).
+                     </t>
+                     <sourcecode type="abnf">
+union-element =/ integer
+integer = [%x2D] (%x30 / (%x31-39 *%x30-39))
+                            ; optional - followed by 0 or
+                            ; sequence of digits with no leading zero
+                     </sourcecode>
+                     <t>Note: the syntax does not allow integers with leading zeros such as <tt>01</tt> and <tt>-01</tt>.
+                     </t>
+                  </section>       
+                  <section numbered="false" toc="exclude">
+                     <name>Semantics</name>
+                     <t>An array index matches any node which is an array and selects any element of the array indexed by the integer index.
```

Comment:
Maybe we should avoid defining index at all and rather consider it as a special case of a slice (when onbly start index is given)?

Reply:
I wanted to follow Proposal A and keep the grammar productions for index and slice separate because that approach neatly avoids some the issues by distinguishing between union elements which can select multiple elements and those which can select at most one element. But we may be able to organise the document to keep these definitions close to each other.